[docs][data][llm] Batch inference docs reorg + update to reflect per-stage config refactor

nrghosh opened the pull request:
## Update / streamline batch inference documentation
- Include new layout of updated architecture
- Reflect refactor for batch inference config knobs / etc
- Simplify user walkthrough, examples, and onboarding for batch inference
- Misc API spec / comment updates and improvements

### New structure for Batch Inference Docs (proposed)

```
Working with LLMs
=================

1. Quickstart   
2. Architecture                       
3. Common use cases
   - Text generation (merge from "Perform batch inference")
   - Embeddings (keep concise)
   - Vision-language models (simplify)
   - OpenAI-compatible endpoints
4. Troubleshooting                     
   - GPU memory / CUDA OOM
   - Model caching for large clusters
5. Advanced configuration            
   - Per-stage configuration
   - Model parallelism (TP/PP)
   - Cross-node parallelism
   - LoRA inference
   - S3/GCS model loading
   - Serve deployments
6. Usage data collection            
```

nrghosh committed:
[docs][data][llm] Update docs for per-stage config refactor

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Merge branch 'master' into data-llm-docs-config-knobs

nrghosh committed:
Merge branch 'master' into data-llm-docs-config-knobs

nrghosh committed:
[docs][data][llm] Reorganize docs for better user onboarding

- Restructure into: Getting Started, Common Use Cases, Troubleshooting, Advanced Config
- Remove redundant 'Perform batch inference' section (duplicated quickstart)
- Promote GPU OOM / model caching to Troubleshooting section
- Consolidate advanced topics (parallelism, per-stage config, LoRA, Serve)
- Simplify VLM and embeddings examples
- Update to new stage config API (prepare_image_stage, etc.)
- Add PIL, RunAI to Vale vocabulary

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Merge branch 'master' into data-llm-docs-config-knobs

nrghosh committed:
Merge branch 'master' into data-llm-docs-config-knobs

Signed-off-by: Nikhil G <nrghosh@users.noreply.github.com>

nrghosh committed:
fix doc

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

richardliaw commented on doc/source/data/working-with-llms.rst:
```diff
@@ -48,305 +57,292 @@ The processor expects input rows with a ``prompt`` field and outputs rows with b
+        --bucket-uri gs://my-bucket/path/to/model
 
-For a complete embedding configuration example, see:
+Then reference the remote path in your config:
```
i would link out to the runai streamer explicitly for further reading

richardliaw replied:
what's the difference between this and below model loading section?

nrghosh replied:
after - (HF model loading) 
<img width="1157" height="542" alt="image" src="https://github.com/user-attachments/assets/ae292e9f-f9f7-492c-9bff-86268ffbfd12" />

richardliaw commented on doc/source/data/working-with-llms.rst:
```diff
@@ -48,305 +57,292 @@ The processor expects input rows with a ``prompt`` field and outputs rows with b
 .. literalinclude:: doc_code/working-with-llms/basic_llm_example.py
     :language: python
     :start-after: __concurrent_config_example_start__
     :end-before: __concurrent_config_example_end__
```
this belongs more in Common rather than Advanced

nrghosh replied:
Common is more for detailing use-cases, I'll add it to 'getting started' - since it applies to all.

nrghosh replied:
after: 
<img width="1130" height="606" alt="image" src="https://github.com/user-attachments/assets/17f8a332-17f6-45fa-a7eb-4c9046e675f8" />

richardliaw commented on doc/source/data/working-with-llms.rst:
```diff
@@ -48,305 +57,292 @@ The processor expects input rows with a ``prompt`` field and outputs rows with b
-    :start-after: __s3_config_example_start__
-    :end-before: __s3_config_example_end__
+    :start-after: __shared_vllm_engine_config_example_start__
+    :end-before: __shared_vllm_engine_config_example_end__
```
this is a pretty important pattern. i would be explicit to use key description words like multi-turn and agentic.

richardliaw reviewed (commented):
overall this change makes sense.

nrghosh committed:
clean up stage diagram

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Address feedback

- move horiz scaling section
- add explicit RunAI streamer link

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
address feedback

- highlight agentic and multi-turn
- add stage configs to api reference
- link to api reference instead of listing fields

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

richardliaw commented on doc/source/data/working-with-llms.rst:
```diff
@@ -3,17 +3,25 @@
 =================
 
-The :ref:`ray.data.llm <llm-ref>` module integrates with key large language model (LLM) inference engines and deployed models to enable LLM batch inference.
+The :ref:`ray.data.llm <llm-ref>` module integrates with LLM inference engines (vLLM, SGLang) to enable scalable batch inference on Ray Data datasets.
```
I think you also need to start with saying that we support both hosted endpoints and running vLLM in Ray Data

richardliaw commented on doc/source/data/working-with-llms.rst:
```diff
@@ -50,40 +58,74 @@ The processor expects input rows with a ``prompt`` field and outputs rows with b
+---------------
 
-Some models may require a Hugging Face token to be specified. You can specify the token in the `runtime_env` argument.
+Use :class:`vLLMEngineProcessorConfig <ray.data.llm.vLLMEngineProcessorConfig>` for chat completions and text generation tasks.
```
can't you use SGLang as well?

jeffreywang-anyscale commented on doc/source/data/working-with-llms.rst:
```diff
@@ -3,17 +3,25 @@
+* :ref:`Architecture <processor_architecture>` - Understand the processor pipeline
+* :ref:`Scaling <horizontal_scaling>` - Scale your LLM stage to multiple replicas
+
+**Common use cases:**
```
I think classification remains one of the most popular use cases, especially with the emergence of Nemo curator. Shall we keep that in the getting started section?

jeffreywang-anyscale commented on doc/source/data/working-with-llms.rst:
```diff
@@ -50,40 +58,74 @@ The processor expects input rows with a ``prompt`` field and outputs rows with b
+         |
+         v
+    - Preprocess (Custom Function)
+    - PrepareMultimodal (Optional, for VLMs)
```
nit: for VLMs and omni models (I don't think the term VLM necessarily includes omni models)

jeffreywang-anyscale commented:
Thanks for updating the doc! It looks much nicer now :)

nrghosh committed:
Address feedback

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
feedback - add note about hosted endpoint / serve deployment

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
feedback - add back classification ex

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
add tpu

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Merge upstream/master

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
fix

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

bveeramani reviewed (approved):
Stamp
